### DISCUSSION TITLE
Make xDrip compatible with latest Full Android Smartwatches

### DISCUSSION THREAD

**Discussion by parapenT1sta** [NightscoutFoundation/xDrip#1973] [General]
As per discussed on this [topic](https://github.com/Navid200/xDrip/discussions/161), xDrip is detecting Full Android Smartwaches as Android Wear (wearOS) Smartwaches and Bluetooth Service stops after smartwatches are wrongly detected.

I have compiled a version with @tolot27 code suggestion and shared on [xDrip Facebook](https://www.facebook.com/groups/xDripG5/posts/6823530591053711/) group and this is the list of watches tested and confirmed the fix:
- LEMFO LEMP
- Rogbid Brave Pro
- Kospet Optimus 2
- Rogbid AIR

The patch applied on `app/src/main/java/com/eveningoutpost/dexdrip/Services/Ob1G5CollectionService.java`:
Line [972](https://github.com/NightscoutFoundation/xDrip/blob/5bf48f918ce10bd59c8eb855551c2ae3d663ded7/app/src/main/java/com/eveningoutpost/dexdrip/Services/Ob1G5CollectionService.java#L972)
**Before:**
`android_wear = (getResources().getConfiguration().uiMode & Configuration.UI_MODE_TYPE_MASK) == Configuration.UI_MODE_TYPE_WATCH;`
**After:**
`android_wear = getPackageManager().hasSystemFeature(PackageManager.FEATURE_WATCH);
`
**Adding at the top of the file:**
`import android.content.pm.PackageManager;`

The patch applied on `app/src/main/java/com/eveningoutpost/dexdrip/Services/DexCollectionService.java`:
Lines [821 - 829](https://github.com/NightscoutFoundation/xDrip/blob/f43d0377467dcbfa968c43a0b776bee838798548/app/src/main/java/com/eveningoutpost/dexdrip/Services/DexCollectionService.java#L821-L829)
**Before:**
```
    @SuppressLint("ObsoleteSdkInt")
    private static boolean shouldServiceRun() {
        if (android.os.Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN_MR2) return false;
        final boolean result = (DexCollectionType.hasXbridgeWixel() || DexCollectionType.hasBtWixel())
                && ((!Home.get_forced_wear() && (((UiModeManager) xdrip.getAppContext().getSystemService(UI_MODE_SERVICE)).getCurrentModeType() != Configuration.UI_MODE_TYPE_WATCH))
                || PersistentStore.getBoolean(CollectionServiceStarter.pref_run_wear_collector));
        if (d) Log.d(TAG, "shouldServiceRun() returning: " + result);
        return result;
    }
```
**After:**
```
    @SuppressLint("ObsoleteSdkInt")
    private boolean shouldServiceRun() {
        if (android.os.Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN_MR2) return false;
        final boolean result = (DexCollectionType.hasXbridgeWixel() || DexCollectionType.hasBtWixel())
                && ((!Home.get_forced_wear() && !getPackageManager().hasSystemFeature(PackageManager.FEATURE_WATCH))
                || PersistentStore.getBoolean(CollectionServiceStarter.pref_run_wear_collector));
        if (d) Log.d(TAG, "shouldServiceRun() returning: " + result);
        return result;
    }
```
**Adding at the top of the file:**
`import android.content.pm.PackageManager;`

Files changed:
- app/src/main/java/com/eveningoutpost/dexdrip/Services/DexCollectionService.java
- app/src/main/java/com/eveningoutpost/dexdrip/Services/Ob1G5CollectionService.java

Files where `Configuration.UI_MODE_TYPE_WATCH` is present:
https://github.com/NightscoutFoundation/xDrip/search?q=Configuration.UI_MODE_TYPE_WATCH

**Comment by Navid200** [comment]
@parapenT1sta Where you asked to open an issue?

**Reply by parapenT1sta** [reply]
Because it's an issue! Smartwaches aren't working. But OK you are the Boss!

**Comment by Navid200** [comment]
The request to not open an issue, unless directed by a developer, which you could have seen when you opened this, applies to all users.
Having an open issue does not expedite solving the issue.  

You already have the attention of tolot27.  He has authority to open issues when/if needed.  
If you focus your energy on what he advises, it's the best you can do to get this resolved.

**Reply by tolot27** [reply]
> You already have the attention of tolot27. He has authority to open issues when/if needed.

@Navid200 If you see that I'm involved, please let me decide if an issue should be created or closed. It is not necessary to close new issues as soon as they have been opened. You can wait at least a few hours up to a couple of days before closing them.

In this particular discussion I already thought about creating an issue by using "Create issue from discussion" on the top right.

**Reply by Navid200** [reply]
@tolot27 Wouldn't it be better to convert the original discussion to an issue in that case rather than creating a new one?
Either way, we have asked all users not to open issues.  

You can convert this back to an issue.  You can convert the original discussion to an issue.  You can ask me to do either one of those two if you prefer.  
Just let me know.

**Reply by parapenT1sta** [reply]
> > You already have the attention of tolot27. He has authority to open issues when/if needed.
> 
> @Navid200 If you see that I'm involved, please let me decide if an issue should be created or closed. It is not necessary to close new issues as soon as they have been opened. You can wait at least a few hours up to a couple of days before closing them.
> 
> In this particular discussion I already thought about creating an issue by using "Create issue from discussion" on the top right.

@tolot27  Thank you for the support! I just provided the resume of what was done but I'm not sure if the same patch can be used in the other files where Configuration.UI_MODE_TYPE_WATCH is present, as wearOS files.

**Reply by tolot27** [reply]
> You can convert this back to an issue. You can convert the original discussion to an issue.

It looks like I don't have the permission to do this in the testing category. Strange. Can you convert [#161 ](https://github.com/Navid200/xDrip/discussions/161) into an issue or assign labels?

**Reply by Navid200** [reply]
That is a discussion in my repository.  Neither of us can move those discussions to the main repository discussions.

Both of us should be able to convert a discussion, in the main repository, into an issue.
Or, did I misunderstand you?

**Reply by tolot27** [reply]
No, you didn't. But it looks like we both have overseen that the original discussion is located in your repository, since you wrote "You can convert the original discussion to an issue." Indeed, none of us can convert those discussions into an issue in another repository.

I'll try to convert this discussion back to its originating issue.

**Reply by tolot27** [reply]
Hmm, a new issue would be created if I convert this discussion back into an issue. Unfortunately, I cannot reopen #1972.

The problem is described partially at https://github.com/github/feedback/discussions/3197.

**Reply by parapenT1sta** [reply]
> Hmm, a new issue would be created if I convert this discussion back into an issue. Unfortunately, I cannot reopen #1972.
> 
> The problem is described partially at [github/feedback#3197](https://github.com/github/feedback/discussions/3197).

If you want I can create a new issue, for me it's OK.

**Comment by tolot27** [comment]
https://github.com/NightscoutFoundation/xDrip/pull/1988 was merged today.

**Comment by Navid200** [comment]
Is any full Android watch still being incorrectly detected as Wear OS by xDrip?
Is this still a problem?

**Reply by parapenT1sta** [reply]
This can be closed. Was solved at the time. Thanks.
